create_layout.py

```diff
@@ -91,34 +91,83 @@
 import in_toto.models.link
 import in_toto.models.layout
 
-def create_material_rules(links, index):
+def snapshot(before_dict, after_dict):
+  '''before_after_snapshot is a simple function that returns which files were
+    unchanged, modified, added or removed from an input dictionary (before_dict)
+    and an output dictionary (after_dict). Both these dictionaries have file
+    names as the keys and their hashes as the values.'''
+
+  unchanged_files = []
+  modified_files = []
+  added_files = []
+  removed_files = []
+  for key in before_dict:
+    if key in after_dict:
+      if before_dict[key] == after_dict[key]:
+        # Matching the hashes to check if file was unchanged
+        unchanged_files.append(key)
+      else:
+        modified_files.append(key)
+    else:
+      removed_files.append(key)
+  for key in after_dict:
+    if key not in before_dict:
+      # Looking for new files
+      added_files.append(key)
+
+  # Returning the snapshot of the new file system
+  return (sorted(unchanged_files), sorted(modified_files), sorted(added_files),
+  sorted(removed_files))
+
+def create_material_rules(current_snapshot, links, index):
   """Create generic material rules (3 variants)
 
   * MATCH available materials with products from previous step (links must be an
   ordered list) and
-  * ALLOW available materials if it is the first step in the
-  list
+  * ALLOW available materials
+  * DELETE removed materials
+  * DISALLOW everything else
+
   Returns a list of material rules
   NOTE: Read header docstring for ideas for more complexity.  """
 
   expected_materials = []
 
-  if index == 0:
-    for material_name in links[index].materials.keys():
-      expected_materials.append(["ALLOW", material_name])
-    expected_materials.append(["DISALLOW", "*"])
+  if index != 0:
+    pre_step_name = links[index-1].name
+    pre_snapshot = snapshot(links[index-1].materials, links[index-1].products)
+
+    for i in range(3):
```

Comment:
Any reason for `3` being hard coded? :)

Reply:
Instead, let's attach each return of `snapshot` to a separate variable?

```
pre_unchanged_files, pre_modified_files, pre_added_files, pre_removed_files = snapshot(...)
```

what do you think?

Reply:
+1 for tuple unpacking!